config/config.go

```diff
@@ -34,38 +39,104 @@ type BitswapConfig struct {
 	TargetMessageSize          int
 }
 
-func (cfg *Config) load(filename string) error {
-	/*
-		f, err := os.Open(filename)
-		if err != nil {
-			if os.IsNotExist(err) {
-				err = ErrNotInitialized
-			}
-			return err
+func (cfg *Config) AddListener(newAddr string) {
+	if !cfg.HasListener(newAddr) {
+		cfg.ListenAddrs = append(cfg.ListenAddrs, newAddr)
+	}
+}
+
+func (cfg *Config) HasListener(find string) bool {
+	for _, addr := range cfg.ListenAddrs {
+		if addr == find {
+			return true
 		}
-		defer f.Close()
-		if err := json.NewDecoder(f).Decode(cfg); err != nil {
-			return fmt.Errorf("failure to decode config: %s", err)
+	}
+	return false
+}
+
+// encode configuration with JSON
+func encode(cfg interface{}, w io.Writer) error {
+	// need to prettyprint, hence MarshalIndent, instead of Encoder
+	buf, err := json.MarshalIndent(cfg, "", "  ")
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(buf)
+	return err
+}
+
+func load(cfg interface{}, filename string) error {
+	f, err := os.Open(filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = ErrNotInitialized
 		}
-	*/
+		return err
+	}
+	defer f.Close()
+	if err := json.NewDecoder(f).Decode(cfg); err != nil {
+		return fmt.Errorf("failure to decode config: %s", err)
+	}
 	return nil
 }
 
-// WriteConfigFile writes the config from `cfg` into `filename`.
-func WriteConfigFile(filename string, cfg interface{}) error {
-	/*
-		err := os.MkdirAll(filepath.Dir(filename), 0755)
-		if err != nil {
-			return err
-		}
+// save writes the config from `cfg` into `filename`.
+func save(cfg interface{}, filename string) error {
+	err := os.MkdirAll(filepath.Dir(filename), 0755)
+	if err != nil {
+		return err
+	}
 
-		f, err := atomicfile.New(filename, 0600)
-		if err != nil {
-			return err
-		}
-		defer f.Close()
+	f, err := atomicfile.New(filename, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
 
-		return encode(f, cfg)
-	*/
-	return nil
+	return encode(cfg, f)
+}
+
+var ErrEmptyPath = errors.New("node not initialized, please run configure")
+
+func MakeAbsolute(root string, path string) (error, string) {
```

Comment:
please change the return type order `(string, error)`

Reply:
Oops. Sorry, I'm a bit of a go newb. I can see how this would irritate!!